core/src/test/java/com/radixdlt/api/core/TransactionPreviewTest.java

```diff
@@ -66,14 +66,154 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+import com.google.common.collect.MoreCollectors;
 import com.radixdlt.api.DeterministicCoreApiTestBase;
+import com.radixdlt.api.core.generated.client.ApiException;
 import com.radixdlt.api.core.generated.models.*;
+import com.radixdlt.crypto.ECKeyPair;
+import com.radixdlt.environment.DatabaseConfig;
+import com.radixdlt.environment.StateHashTreeGcConfig;
+import com.radixdlt.harness.deterministic.DeterministicTest;
+import com.radixdlt.harness.predicates.NodesPredicate;
+import com.radixdlt.identifiers.Address;
+import com.radixdlt.lang.Functions;
 import com.radixdlt.rev2.Manifest;
+import com.radixdlt.testutil.TestStateReader;
 import com.radixdlt.utils.Bytes;
 import com.radixdlt.utils.PrivateKeys;
+import com.radixdlt.utils.UInt32;
+import com.radixdlt.utils.UInt64;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.awaitility.Awaitility;
 import org.junit.Test;
 
 public class TransactionPreviewTest extends DeterministicCoreApiTestBase {
+
+  // a known amount coming from faucet; for assert purposes
+  private static final double FAUCET_AMOUNT = 10000;
+
+  @Test
+  public void transaction_preview_executes_at_historical_version() throws Exception {
+    try (var test = buildTest(true, 20L)) {
+      test.suppressUnusedWarning();
+
+      // Prepare a simple manifest which just gets some amount from faucet, once:
+      var accountKeyPair = ECKeyPair.generateNew();
+      var accountAddress = Address.virtualAccountAddress(accountKeyPair.getPublicKey());
+      var manifest = Manifest.depositFromFaucet(accountAddress);
+
+      // Execute it once, to initialize the account and learn its XRD vault address:
+      submitAndWaitForSuccess(test, manifest, List.of(accountKeyPair));
+      var initialVaultBalance =
+          this.getStateApi()
+              .stateAccountPost(
+                  new StateAccountRequest()
+                      .network(networkLogicalName)
+                      .accountAddress(addressing.encode(accountAddress)))
+              .getVaults()
+              .stream()
+              .collect(MoreCollectors.onlyElement());
+      var vaultAddress = initialVaultBalance.getVaultEntity().getEntityAddress();
+
+      // Sanity check - the vault has a known balance (the "1x from Faucet" amount):
+      var initialAmount = (FungibleResourceAmount) initialVaultBalance.getResourceAmount();
+      assertThat(Double.parseDouble(initialAmount.getAmount())).isEqualTo(FAUCET_AMOUNT);
+
+      // Reach a known state version:
+      test.runUntilState(NodesPredicate.anyAtOrOverStateVersion(37));
```

Comment:
Hmm what's the meaning of `37`? Can we instead just read the state version of a node which has committed the above transaction from the API?

Or use `var commitResult = submitAndWaitForSuccess(test, manifest, List.of(accountKeyPair)); commitResult.stateVersion`?

Reply:
> Hmm what's the meaning of 37?

it could also be 42 or 1337; but you are right, better to use exact, captured numbers 👍 PTAL